realm/realm-library/src/main/cpp/io_realm_internal_OsRealmConfig.cpp

```diff
@@ -142,18 +142,22 @@ JNIEXPORT void JNICALL Java_io_realm_internal_OsRealmConfig_nativeSetSchemaConfi
             config.migration_function = [j_migration_cb_weak, j_config_weak](SharedRealm old_realm,
                                                                                  SharedRealm realm, Schema&) {
                 JNIEnv* env = JniUtils::get_env(false);
+                // Java needs a new pointer for the SharedRealm life control.
+                SharedRealm* new_shared_realm_ptr = new SharedRealm(realm);
                 JavaGlobalRef config_global = j_config_weak.global_ref(env);
                 if (!config_global) {
                     return;
                 }
 
                 j_migration_cb_weak.call_with_local_ref(env, [&](JNIEnv* env, jobject obj) {
-                    // Java doesn't own the &realm pointer!
                     env->CallStaticVoidMethod(get_shared_realm_class(env), run_migration_callback_method,
-                                              reinterpret_cast<jlong>(&realm), config_global.get(), obj,
+                                              reinterpret_cast<jlong>(new_shared_realm_ptr), config_global.get(), obj,
                                               old_realm->schema_version());
                 });
-                TERMINATE_JNI_IF_JAVA_EXCEPTION_OCCURRED(env);
+                // Close the SharedRealm. Otherwise it will only be closed when the Java OsSharedRealm gets GCed. And
+                // that will be too late.
+                TERMINATE_JNI_IF_JAVA_EXCEPTION_OCCURRED(
+                    env, [&new_shared_realm_ptr]() { (*new_shared_realm_ptr)->close(); });
```

Comment:
do we need to `delete new_shared_realm_ptr` as well?

Reply:
nope~ it will be still handled by the GC and phantom refs.